bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogContext.java

```diff
@@ -95,72 +99,112 @@ public Locale locale() {
+        private List<HumanLanguageInterpreter> hlis = List.of();
+        // options
+        private @Nullable String listeningItem;
+        private String keyword;
+        private Locale locale;
+
+        public Builder(String keyword, Locale locale) {
+            this.keyword = keyword;
+            this.locale = locale;
+        }
+
+        public Builder withSource(@Nullable AudioSource source) {
+            this.source = source;
+            return this;
+        }
+
+        public Builder withSink(@Nullable AudioSink sink) {
+            this.sink = sink;
+            return this;
+        }
+
+        public Builder withKS(@Nullable KSService service) {
+            this.ks = service;
+            return this;
+        }
+
+        public Builder withSTT(@Nullable STTService service) {
+            this.stt = service;
+            return this;
+        }
+
+        public Builder withTTS(@Nullable TTSService service) {
+            this.tts = service;
+            return this;
+        }
+
+        public Builder withHLIs(Collection<HumanLanguageInterpreter> services) {
+            return withHLIs(new ArrayList<>(services));
+        }
+
+        public Builder withHLIs(List<HumanLanguageInterpreter> services) {
+            this.hlis = services;
+            return this;
+        }
+
+        public Builder withKeyword(String keyword) {
+            this.keyword = keyword;
+            return this;
+        }
+
+        public Builder withVoice(@Nullable Voice voice) {
+            this.voice = voice;
+            return this;
+        }
+
+        public Builder withListeningItem(@Nullable String listeningItem) {
+            this.listeningItem = listeningItem;
+            return this;
+        }
+
+        public Builder withLocale(Locale locale) {
+            this.locale = locale;
+            return this;
+        }
+
+        public DialogContext build() throws IllegalStateException {
+            KSService ksService = ks;
+            STTService sttService = stt;
+            TTSService ttsService = tts;
+            List<HumanLanguageInterpreter> hliServices = hlis;
+            AudioSource audioSource = source;
+            AudioSink audioSink = sink;
+            List<String> errors = new ArrayList<>();
+            if (sttService == null) {
+                errors.add("Missing stt service");
+            }
+            if (ttsService == null) {
+                errors.add("Missing tts service");
+            }
+            if (hliServices.isEmpty()) {
+                errors.add("Missing interpreters");
+            }
+            if (audioSource == null) {
+                errors.add("Missing audio source");
+            }
+            if (audioSink == null) {
+                errors.add("Missing audio sink");
+            }
+            if (!errors.isEmpty()) {
+                errors.forEach(logger::warn);
+                throw new IllegalStateException("Cannot build dialog context, services are missing");
```

Comment:
Can you add the missing parameters to the `IllegalStateException` message instead (e.g. using `String.join(..)` ?

When exceptions are thrown it should be up to the method caller what to do with them and determine if logging should occur.

You might also want to add a JavaDoc to the `.build()` method to explain when the exception occurs so devs do not have to inspect to code to see when it is thrown.